### DISCUSSION TITLE
Difference for a nested attitbute with in an Entity

### DISCUSSION THREAD

**Discussion by vikramp** [javers/javers#1164] [Q&A]
Hi, 
I am evaluating JAVERS as entity versioning library. I was interested in knowing if there is a way to customize the diff evaluation where one of the Entity attributes is a String containing key - value pairs. See example below

class Employee{ Long id ; String properties;} and properties contains lets says .....{'salary':1,'daysSinceAppraisal':25}"

Of course as is Javers does identity that properties attribute differ between two versions.
I was hoping if I could extend to find out lets say if only salary has changed.

**Comment by bartoszwalacik** [comment]
Javers can't parse your Strings, but you can create a getter returning a Map created from this String and compare these getters

**Reply by vikramp** [reply]
Do I need to do a special configuration to include map in diff evaluation ?
@Builder
@Entity
public class Station {

    @Id
    private Long id;

    private String name;

    @DiffIgnore
    private String nestedJson;

    private Map<String,String> map;

    public void setNestedJson(String nestedJson){
        this.nestedJson = nestedJson;
        this.map = new Gson().fromJson(nestedJson,Map.class);
    }

**Reply by bartoszwalacik** [reply]
no, it works out-of-the-box

**Reply by vikramp** [reply]
Something seems a bit off. 
See below in the output below map attribute isnt counted in the diff.

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import lombok.Builder;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;
import org.javers.core.metamodel.annotation.DiffIgnore;
import org.javers.core.metamodel.annotation.DiffInclude;
import org.javers.core.metamodel.annotation.Entity;
import org.javers.core.metamodel.annotation.Id;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import java.util.HashMap;
import java.util.Map;

@Getter
@Setter
@ToString
@Builder
@Entity
public class Station {

    private static Logger logger = LoggerFactory.getLogger(Station.class);

    @Id
    private Long id;

    private String name;

    @DiffIgnore
    private String nestedJson;

    private Map<String,String> map = new HashMap<String,String>();

    public String getNestedJson() {
        return nestedJson;
    }

    public Map<String, String> getMap() {
        logger.info("Nested Json is  :" + this.nestedJson);
        return new Gson().fromJson(this.nestedJson,Map.class);
    }

}

.........
import Station;
import org.javers.core.Javers;
import org.javers.core.JaversBuilder;
import org.javers.core.MappingStyle;
import org.javers.core.diff.Diff;
import org.junit.jupiter.api.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import java.util.Collections;


public class GraphQLJunitTest {

    private static Logger logger = LoggerFactory.getLogger(GraphQLJunitTest.class);
    @Test
    public void testTestJaVersDetection(){
        Javers javers =  JaversBuilder.javers().withMappingStyle(MappingStyle.FIELD).build();

        Station stationOriginal = Station.builder().id(1L).name("ABCD").nestedJson("{'numberOfTanks':1,'daysSinceInspection':25}").build();
        Station stationCompareTO = Station.builder().id(1L).name("XYZ").nestedJson("{'numberOfTanks':1,'daysSinceInspection':25}").build();
        Diff diff = javers.compare(stationCompareTO,stationOriginal);
        logger.info(diff.prettyPrint());
        Station stationCompareTONested = Station.builder().id(1L).name("DEF").nestedJson("{'numberOfTanks':1,'daysSinceInspection':30}").build();
        diff = javers.compare(stationCompareTONested,stationOriginal);
        logger.info(stationCompareTONested.toString());
        logger.info(stationCompareTO.toString());
        logger.info(diff.prettyPrint());
    }
}
........... ---Output of Junit ------
09:02:42.070 [main] DEBUG org.javers.core.JaversBuilder - starting up JaVers ...
09:02:42.133 [main] INFO org.javers.core.JaversBuilder - mappingStyle: FIELD
09:02:42.136 [main] INFO org.javers.core.JaversBuilder - loading GuavaAddOns ...
09:02:42.200 [main] INFO org.javers.core.JaversBuilder - using fake InMemoryRepository, register actual Repository implementation via JaversBuilder.registerJaversRepository()
09:02:42.219 [main] INFO org.javers.core.JaversBuilder - JaVers instance started in 150 ms
09:02:42.234 [main] DEBUG org.javers.TypeMapper - javersType of 'Station' inferred from annotations as EntityType
09:02:42.239 [main] DEBUG org.javers.TypeMapper - javersType of 'Map' spawned as MapType from prototype MapType{ baseType: 'interface java.util.Map' }
09:02:42.239 [main] DEBUG org.javers.core.graph.ObjectGraphBuilder - live graph assembled, object nodes: 1, entities: 1, valueObjects: 0
09:02:42.240 [main] DEBUG org.javers.core.graph.ObjectGraphBuilder - live graph assembled, object nodes: 1, entities: 1, valueObjects: 0
09:02:42.251 [main] INFO XXXXX.GraphQLJunitTest - Diff:
* changes on Station/1 :
  - 'name' value changed from 'XYZ' to 'ABCD'

09:02:42.252 [main] DEBUG org.javers.core.graph.ObjectGraphBuilder - live graph assembled, object nodes: 1, entities: 1, valueObjects: 0
09:02:42.252 [main] DEBUG org.javers.core.graph.ObjectGraphBuilder - live graph assembled, object nodes: 1, entities: 1, valueObjects: 0
09:02:42.252 [main] INFO XXX.Station - Nested Json is  :{'numberOfTanks':1,'daysSinceInspection':30}
09:02:42.258 [main] INFO XXX.GraphQLJunitTest - Station(id=1, name=DEF, nestedJson={'numberOfTanks':1,'daysSinceInspection':30}, map={numberOfTanks=1.0, daysSinceInspection=30.0})
09:02:42.258 [main] INFO XX.Station - Nested Json is  :{'numberOfTanks':1,'daysSinceInspection':25}
09:02:42.258 [main] INFO XX.GraphQLJunitTest - Station(id=1, name=XYZ, nestedJson={'numberOfTanks':1,'daysSinceInspection':25}, map={numberOfTanks=1.0, daysSinceInspection=25.0})
09:02:42.258 [main] INFO XXX.GraphQLJunitTest - Diff:
* changes on XXX.Station/1 :
  - 'name' value changed from 'DEF' to 'ABCD'


Process finished with exit code 0

**Reply by vikramp** [reply]
Let me know if I missed out in the original post. A bit of changes in the way Lombok was configured appeared to give me the desired results. See below

@Builder(builderClassName = "StationBuilder")
@Data
@Entity
@AllArgsConstructor
@NoArgsConstructor
public class Station {

    private static Logger logger = LoggerFactory.getLogger(Station.class);

    @Id
    private Long id;
    private String name;


    @DiffIgnore
    private String nestedJson;
    private Map<String,String> propertiesMap;


    public static class StationBuilder{

        private String nestedJson ;
        private Map<String,String> propertiesMap;

        public StationBuilder nestedJson(String nestJson) {
            this.nestedJson = nestJson;
            this.propertiesMap = new Gson().fromJson(nestJson,Map.class);
            return this;
        }
    }
}
-------- Output ------
* changes on XX.Station/1:
  - 'name' value changed from 'DEF' to 'ABCD'
  - 'propertiesMap' map changes :
    'daysSinceInspection' -> '30.0' changed to '25.0'

**Comment by vikramp** [comment]
Appears a bit complicated, Is there another way to simplify this ?

**Reply by bartoszwalacik** [reply]
I'm not sure what's your question, if you are asking about Lombok - i'm not a fan.
